### ISSUE TITLE
Enhancement: [no-explicit-any] suggest to replace keyof any with PropertyKey

### ISSUE THREAD

**Issue by KuSh** [typescript-eslint/typescript-eslint#11009]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-explicit-any

### Description

Suggest to replace keyof any (and keyof never?) with PropertyKey as it is more explicit.
This can even be an auto fixable thing.

### Fail

```TypeScript
const integer = <TKey extends keyof any, TTarget extends { [K in TKey]: number }>(
  target: TTarget,
  key: TKey
) => { /* ... */ };
```

### Pass

```TypeScript
const integer = <TKey extends PropertyKey, TTarget extends { [K in TKey]: number }>(
  target: TTarget,
  key: TKey
) => { /* ... */ };
```

### Additional Info

Extracted from https://github.com/typescript-eslint/typescript-eslint/issues/3206

**Comment by JoshuaKGoldberg** [comment]
Interesting idea! My vote is 👍, this would be a nice improvement to the rule. Knowing that `keyof any` gives `string | number | symbol` / `PropertyKey` is a bit niche of knowledge. I'd wager that some non-trivial slice of users have copy & pasted it then gotten confused about the rule reporting.
